NetKet/Graph/lattice_impl.hpp

```diff
@@ -0,0 +1,430 @@
+  std::vector<int> colors(Nsites(), -1);
+  const auto adjacency_list =
+      AdjacencyList();  // implicit expression can't have
+                        // access to the Lattice function
+  if (is_connected_) {
+    BreadthFirstSearch(
+        start, [&colors, &adjacency_list, &is_bipartite](int node, int) {
+          if (node == start) colors[node] = 1;
+          for (std::size_t j = 0; j < adjacency_list[node].size(); j++) {
+            if (!is_bipartite) break;
+            if (colors[adjacency_list[node][j]] == -1) {
+              colors[adjacency_list[node][j]] = 1 - colors[node];
+            } else if (colors[adjacency_list[node][j]] == colors[node]) {
+              is_bipartite = false;
+            }
+          }
+        });
+  } else {
+    BreadthFirstSearch(
+        [&colors, &adjacency_list, &is_bipartite](int node, int, int) {
+          if (node == start) colors[node] = 1;
+          for (std::size_t j = 0; j < adjacency_list[node].size(); j++) {
+            if (!is_bipartite) break;
+            if (colors[adjacency_list[node][j]] == -1) {
+              colors[adjacency_list[node][j]] = 1 - colors[node];
+            } else if (colors[adjacency_list[node][j]] == colors[node]) {
+              is_bipartite = false;
+            }
+          }
+        });
+  }
+  return is_bipartite;
+}
+
+// Lattice representations (site = k, vector = n_i, coord = cartesian
+// coordinates)
+std::vector<int> Lattice::Site2Vector(int i) const {
+  int ndim = extent_.size();
+  std::vector<int> result(ndim, 0);
+  int ip;
+  if (i < nlatticesites_)
+    ip = i;
+  else
+    ip = i % nlatticesites_;
+  int k = ndim - 1;
+  while (ip > 0) {
+    result[k] = ip % extent_[k];
+    ip /= extent_[k];
+    k--;
+  }
+  return result;
+};
+std::vector<double> Lattice::Vector2Coord(const std::vector<int> &n,
+                                          int iatom) const {
+  std::vector<double> R(ndim_, 0);
+  for (int j = 0; j < ndim_; j++) {
+    R[j] = atoms_coord_[iatom][j];
+    for (int k = 0; k < ndim_; k++) R[j] += n[k] * basis_vectors_[k][j];
+  }
+
+  return R;
+};
+std::vector<double> Lattice::Site2Coord(int k) const { return R_[k]; }
+int Lattice::Vector2Site(const std::vector<int> &n) const {
+  int k = 0;
+
+  for (int i = 0; i < Ndim(); i++) {
+    int base = 1;
+    for (int j = i + 1; j < Ndim(); j++) base *= extent_[j];
+    k += n[i] * base;
+  }
+
+  return k;
+}
+
+// Neighbours
+std::vector<std::vector<int>> Lattice::PossibleLatticeNeighbours() const {
+  const int nsymbols = 3;
+  const int max = nsymbols - 1;
+  std::vector<int> b(ndim_, 0);
+  std::vector<int> result_vector;
+  std::vector<std::vector<int>> result_matrix;
+  std::vector<int> row(ndim_, 0);
+  do {
+    for (std::vector<int>::const_iterator ib = b.begin(); ib != b.end(); ++ib)
+      result_vector.push_back(*ib);
+  } while (next_variation(b.begin(), b.end(), max));
+
+  for (std::size_t i = 0; i < result_vector.size() / ndim_; i++) {

```

Comment:
Any chance you could use Eigen for matrix-related stuff? Your re-implementing slicing and copying makes the code a bit difficult to check (at least for me).